src/selectors/per-thread/thread.js

```diff
@@ -0,0 +1,212 @@
+import type { JsTracerTiming } from '../../types/profile-derived';
+import type { $ReturnType } from '../../types/utils';
+
+const _getDefaultCategoryWrappedInObject = createSelector(
+  ProfileSelectors.getDefaultCategory,
+  defaultCategory => ({ value: defaultCategory })
+);
+
+/**
+ * Infer the return type from the getThreadSelectorsPerThread function. This
+ * is done that so that the local type definition with `Selector<T>` is the canonical
+ * definition for the type of the selector.
+ */
+export type ThreadSelectorsPerThread = $ReturnType<
+  typeof getThreadSelectorsPerThread
+>;
+
+/**
+ * Create the selectors for a thread that have to do with an entire thread. This includes
+ * the general filtering pipeline for threads.
+ */
+export function getThreadSelectorsPerThread(threadIndex: ThreadIndex): * {
+  const getThread: Selector<Thread> = state =>
+    ProfileSelectors.getProfile(state).threads[threadIndex];
+  const getStringTable: Selector<UniqueStringArray> = state =>
+    getThread(state).stringTable;
+
+  /**
+   * The first per-thread selectors filter out and transform a thread based on user's
+   * interactions. The transforms are order dependendent.
+   *
+   * 1. Unfiltered getThread - The first selector gets the unmodified original thread.
+   * 2. Range - New samples table with only samples in the committed range.
+   * 3. Transform - Apply the transform stack that modifies the stacks and samples.
+   * 4. Implementation - Modify stacks and samples to only show a single implementation.
+   * 5. Search - Exclude samples that don't include some text in the stack.
+   * 6. Preview - Only include samples that are within a user's preview range selection.
+   */
+  const getRangeFilteredThread: Selector<Thread> = createSelector(
+    getThread,
+    ProfileSelectors.getCommittedRange,
+    (thread, range) => {
+      const { start, end } = range;
+      return ProfileData.filterThreadToRange(thread, start, end);
+    }
+  );
+
+  // It becomes very expensive to apply each transform over and over again as they
+  // typically take around 100ms to run per transform on a fast machine. Memoize
+  // memoize each step individually so that they transform stack can be pushed and
+  // popped frequently and easily.
+  const applyTransformMemoized = memoize(Transforms.applyTransform, {
+    cache: new MixedTupleMap(),
+  });
+
+  const getTransformStack: Selector<TransformStack> = state =>
+    UrlState.getTransformStack(state, threadIndex);
+
+  const getRangeAndTransformFilteredThread: Selector<Thread> = createSelector(
+    getRangeFilteredThread,
+    getTransformStack,
+    _getDefaultCategoryWrappedInObject,
+    (startingThread, transforms, defaultCategoryObj) =>
+      transforms.reduce(
+        // Apply the reducer using an arrow function to ensure correct memoization.
+        (thread, transform) =>
+          applyTransformMemoized(thread, transform, defaultCategoryObj.value),
```

Comment:
ah looks like you'll have a conflict here :)

Reply:
Splitting apart files is really bad for merging in upstream work.